File: leetcode-solutions/dp/solution-0087/solution.js
/**
 * @param {string} s1
 * @param {string} s2
 * @return {boolean}
 */
let dp = {};
var isScramble = function(s1, s2) {
   if (s1 === s2) return true;
   let key = s1 + ' ' + s2;
   if (dp[key] !== undefined) return dp[key];
   for (let i = 0; i < s1.length - 1; i++) {
     let a = isScramble(s1.slice(0, i + 1), s2.slice(0, i + 1));
     let b = isScramble(s1.slice(i + 1), s2.slice(i + 1));
     let c = isScramble(s1.slice(0, i + 1), s2.slice(s2.length - 1 - i));
     let d = isScramble(s1.slice(i + 1), s2.slice(0, s2.length - 1 - i));
     if (a && b || c && d) return dp[key] = true;
   }
   return dp[key] = false;
 };
 

File: leetcode-solutions/dp/solution-0087/solution.go
package solution0087

var dp = make(map[string]bool)

func isScramble(s1 string, s2 string) bool {
    key := s1 + s2
    if v, ok := dp[key]; ok {
        return v
    }
    if s1 == s2 {
        return true
    }
    for i := 0; i < len(s1)-1; i++ {
        a := isScramble(s1[:i+1], s2[:i+1])
        b := isScramble(s1[i+1:], s2[i+1:])
        c := isScramble(s1[:i+1], s2[len(s2)-1-i:])
        d := isScramble(s1[i+1:], s2[:len(s2)-1-i])
        if a && b || c && d {
            dp[key] = true
            return true
        }
    }
    dp[key] = false
    return false
}
